components/places/src/storage/bookmarks.rs

```diff
@@ -376,6 +395,259 @@ fn delete_bookmark_in_tx(db: &impl ConnExt, guid: &SyncGuid) -> Result<bool> {
+            position = existing.position;
+            update_old_parent_status = false;
+            update_new_parent_status = false;
+        }
+        UpdateTreeLocation::Position(pos) => {
+            parent_id = existing.parent_id;
+            update_old_parent_status = true;
+            update_new_parent_status = false;
+            // Not clear that InvalidParent is the correct error here - probably
+            // should be a "stuff is corrupt" error? Or maybe we should fix it
+            // here by reparenting to unfiled?
+            let parent = get_raw_bookmark(db, &existing.parent_guid)?
+                .ok_or_else(|| InvalidPlaceInfo::InvalidParent(existing.parent_guid.to_string()))?;
+            update_pos_for_deletion(db, existing.position, parent.row_id)?;
+            // We just removed a child, so actual child count is now parent.child_count - 1
+            position = resolve_pos_for_insert(db, pos, parent.row_id, parent.child_count - 1)?;
+        }
+        UpdateTreeLocation::Parent(new_parent_guid, pos) => {
+            if BookmarkRootGuid::from_guid(&new_parent_guid) == Some(BookmarkRootGuid::Root) {
+                return Err(InvalidPlaceInfo::InvalidGuid.into());
+            }
+            let new_parent = get_raw_bookmark(db, &new_parent_guid)?
+                .ok_or_else(|| InvalidPlaceInfo::InvalidParent(new_parent_guid.to_string()))?;
+            if new_parent.bookmark_type != BookmarkType::Folder {
+                return Err(InvalidPlaceInfo::InvalidParent(new_parent_guid.to_string()).into());
+            }
+            parent_id = new_parent.row_id;
+            update_old_parent_status = true;
+            update_new_parent_status = true;
+            // This position change is more complicated across parents.
+            // As above, this failure really means "stuff is corrupt" and we
+            // could consider fixing the tree instead of throwing?
+            let existing_parent = get_raw_bookmark(db, &existing.parent_guid)?
+                .ok_or_else(|| InvalidPlaceInfo::InvalidParent(existing.parent_guid.to_string()))?;
+            update_pos_for_deletion(db, existing.position, existing_parent.row_id)?;
+            position = resolve_pos_for_insert(db, pos, new_parent.row_id, new_parent.child_count)?;
+        }
+    };
+    let place_id = match item {
+        UpdatableItem::Bookmark(b) => match &b.url {
+            None => existing.place_id,
+            Some(url) => {
+                let page_info = match fetch_page_info(db, &url)? {
+                    Some(info) => info.page,
+                    None => new_page_info(db, &url, None)?,
+                };
+                Some(page_info.row_id)
+            }
+        },
+        _ => {
+            // Updating a non-bookmark item, so the existing item must not
+            // have a place_id
+            assert_eq!(existing.place_id, None);
+            None
+        }
+    };
+    let update_title = match item {
```

Comment:
We should truncate titles here the way we do for insertions.

Reply:
Nice catch, done!